### PR TITLE
*: directly save prepare execute args as datums in binary protocol

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

This is 1 of 2 prepare improve PRs

simplify 

```
rawValue -> ast.ExprNode ->  Expression -> Datum
```

to 

```
rawValue -> Datum
```

for binary protocol execute use shorter code path, and for `execute stmt1 using @x` keep old code path.

by the way, this pr also want to cut off this:

image

because using `@x` need exp rewrite but binary execute  theoretically should never do that

### What is changed and how it works?

- change `parseStmtArgs` to  return Datum directly
- execute using Datum created by `parseStmtArgs`

### Check List 

Tests 

before this PR:

 - Unit test
 - Integration test

Code changes

 - implements

Side effects

 - N/A

Related changes

 - do more test then cherry-pick to 3.0

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/10884)

### CODE REVIEW

**Comment by lysu** [comment]
tested plus with https://github.com/pingcap/tidb/pull/10746

before patch

image

image

after this:

image


image

**Comment by lysu** [comment]
@jackysp @tiancaiamao  PTAL if free thx

File: server/conn_stmt.go [hunk 1]
```diff
@@ -38,6 +38,7 @@ import (
 	"context"
 	"encoding/binary"
 	"fmt"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
```
Comment (tiancaiamao) [inline server/conn_stmt.go]: Format

File: session/tidb_test.go [hunk 2]
```diff
@@ -32,6 +24,13 @@ import (
 	"github.com/pingcap/tidb/util/logutil"
 	"github.com/pingcap/tidb/util/sqlexec"
 	"github.com/pingcap/tidb/util/testleak"
+
+	"context"
```
Comment (tiancaiamao) [inline session/tidb_test.go]: ??

File: util/kvencoder/kv_encoder.go [hunk 3]
```diff
@@ -17,6 +17,7 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"github.com/pingcap/tidb/types"
```
Comment (tiancaiamao) [inline util/kvencoder/kv_encoder.go]: ditto

**Pull request merged**